language/move-lang/src/lib.rs

```diff
@@ -716,39 +703,51 @@ fn parse_program(
     }
 
     let res = if errors.is_empty() {
-        Ok((
-            parser::ast::Program {
-                source_definitions,
-                lib_definitions,
-            },
-            source_comments,
-        ))
+        let pprog = parser::ast::Program {
+            source_definitions,
+            lib_definitions,
+        };
+        let pprog = if sources_shadow_deps {
+            shadow_lib_module_definitions(pprog)
+        } else {
+            pprog
+        };
+        Ok((pprog, source_comments))
     } else {
         Err(errors)
     };
     Ok((files, res))
 }
 
-fn check_targets_deps_dont_intersect(
+fn ensure_targets_deps_dont_intersect(
     targets: &[&'static str],
-    deps: &[&'static str],
+    deps: &mut Vec<&'static str>,
+    sources_shadow_deps: bool,
 ) -> anyhow::Result<()> {
     let target_set = targets.iter().collect::<BTreeSet<_>>();
     let dep_set = deps.iter().collect::<BTreeSet<_>>();
-    let intersection = target_set.intersection(&dep_set).collect::<Vec<_>>();
+    let intersection = target_set
+        .intersection(&dep_set)
+        .cloned()
+        .cloned()
+        .collect::<Vec<_>>();
     if intersection.is_empty() {
         return Ok(());
     }
-
-    let all_files = intersection
-        .into_iter()
-        .map(|s| format!("    {}", s))
-        .collect::<Vec<_>>()
-        .join("\n");
-    Err(anyhow!(
-        "The following files were marked as both targets and dependencies:\n{}",
-        all_files
-    ))
+    if sources_shadow_deps {
+        deps.retain(|fname| !intersection.contains(fname));
+        Ok(())
```

Comment:
nit, I would keep this file in guarded style, so `return Ok(())` and drop the `else`